xdsl/parser.py

```diff
@@ -464,39 +466,41 @@ def parse_list(self,
     V = TypeVar('V')
 
     def parse_dictionary(self,
-                         parse_optional_key: Callable[[], K | None],
-                         parse_optional_value: Callable[[], V | None],
+                         parse_key: Callable[[], K],
+                         parse_value: Callable[[], V],
                          delimiter: str = ",",
                          skip_white_space: bool = True) -> dict[K, V]:
         if skip_white_space:
             self.skip_white_space()
         assert (len(delimiter) <= 1)
-        entry = self.parse_optional_dict_entry(parse_optional_key,
-                                               parse_optional_value)
-        if entry is None:
+        if len(delimiter):
+            parse_delimiter = lambda: self.parse_char(delimiter)
+        else:
+            parse_delimiter = lambda: True
+
+        self.parse_char("{")
+        if self.peek_char("}"):
             return {}
-        res = dict([entry])
-        while (self.parse_optional_char(delimiter)
-               if len(delimiter) == 1 else True):
-            entry = self.parse_optional_dict_entry(parse_optional_key,
-                                                   parse_optional_value)
-            if entry is not None:
-                res[entry[0]] = entry[1]
+
+        entry = self.parse_dict_entry(parse_key, parse_value)
+        res = dict((entry, ))
```

Comment:
I am not sure about construction of dictionaries in python, but this looks extremely weird. I understood the replaced code, in what sense is the new version different?

Reply:
Yeah this is probably worth a comment. This is the constructor from an iterable of pairs, and I wrote it this way because the types of the key and value aren't expressible in this context, so no explicit type annotations would make the compiler happy.

Reply:
Turns out the normal syntax works, not sure how it's different to what I tried earlier